rust-runtime/aws-smithy-http-server/src/plugin/pipeline.rs

```diff
@@ -0,0 +1,143 @@
+/// You can use the [`map`](PluginPipeline::map) method to grab the current pipeline and transform it:
+///
+/// ```rust
+/// use aws_smithy_http_server::plugin::{filter_by_operation_name, PluginPipeline};
+/// # use aws_smithy_http_server::plugin::IdentityPlugin as LoggingPlugin;
+/// # use aws_smithy_http_server::plugin::IdentityPlugin as MetricsPlugin;
+/// # use aws_smithy_http_server::plugin::IdentityPlugin as AuthPlugin;
+/// # struct CheckHealth;
+/// # impl CheckHealth { const NAME: &'static str = "MyName"; }
+///
+/// let pipeline = PluginPipeline::new(LoggingPlugin)
+///     .push(MetricsPlugin)
+///     .map(|current_pipeline| {
+///         // The logging and metrics plugins will not be applied to the `CheckHealth` operation.
+///         filter_by_operation_name(current_pipeline, |name| name != CheckHealth::NAME)
+///     })
+///     // The auth plugin will be applied to all operations
+///     .push(AuthPlugin);
+/// ```
+///
+/// ## Concatenating two plugin pipelines
+///
+/// `PluginPipeline` is a good way to bundle together multiple plugins, ensuring they are all
+/// registered in the correct order.
+///
+/// You can use the [`concat`](PluginPipeline::concat) to append, at once, all the plugins
+/// in another pipeline to the current pipeline:
+///
+/// ```rust
+/// use aws_smithy_http_server::plugin::{PluginPipeline, PluginStack};
+/// # use aws_smithy_http_server::plugin::IdentityPlugin as LoggingPlugin;
+/// # use aws_smithy_http_server::plugin::IdentityPlugin as MetricsPlugin;
+/// # use aws_smithy_http_server::plugin::IdentityPlugin as AuthPlugin;
+///
+/// pub fn get_bundled_pipeline() -> PluginPipeline<PluginStack<LoggingPlugin, MetricsPlugin>> {
+///     PluginPipeline::new(LoggingPlugin).push(MetricsPlugin)
+/// }
+///
+/// let pipeline = PluginPipeline::new(AuthPlugin)
+///     .append(get_bundled_pipeline());
+/// ```
+///
+/// ## Providing custom methods on `PluginPipeline`
+///
+/// You use an **extension trait** to add custom methods on `PluginPipeline`.
+///
+/// This is a simple example using `AuthPlugin`:
+///
+/// ```rust
+/// use aws_smithy_http_server::plugin::{PluginPipeline, PluginStack};
+/// # use aws_smithy_http_server::plugin::IdentityPlugin as LoggingPlugin;
+/// # use aws_smithy_http_server::plugin::IdentityPlugin as AuthPlugin;
+///
+/// pub trait AuthPluginExt<CurrentPlugins> {
+///     fn with_auth(self) -> PluginPipeline<PluginStack<CurrentPlugins, AuthPlugin>>;
+/// }
+///
+/// impl<CurrentPlugins> AuthPluginExt<CurrentPlugins> for PluginPipeline<CurrentPlugins> {
+///     fn with_auth(self) -> PluginPipeline<PluginStack<CurrentPlugins, AuthPlugin>> {
+///         self.push(AuthPlugin)
+///     }
+/// }
+///
+/// let pipeline = PluginPipeline::new(LoggingPlugin)
+///     // Our custom method!
+///     .with_auth();
+/// ```
+pub struct PluginPipeline<P>(P);
+
+impl PluginPipeline<IdentityPlugin> {
+    pub fn empty() -> Self {
```

Comment:
That won't work unfortunately:

```
error[E0282]: type annotations needed
 --> src/plugin/pipeline.rs:23:16
  |
8 | let pipeline = PluginPipeline::empty().push(LoggingPlugin).push(MetricsPlugin);
  |                ^^^^^^^^^^^^^^^^^^^^^ cannot infer type for type parameter `P`
```

even if `P` is not used, the compiler must figure out what its value should be and it can't.

Reply:
I see, thanks for checking!